### ISSUE TITLE
zoom does not work on JavaScript buffer

### ISSUE THREAD

**Issue by stardiviner** [cyrus-and/zoom#30]
As this screenshot shows:

![image](https://user-images.githubusercontent.com/378699/85814269-26aceb00-b798-11ea-8d3a-33a1246e8cd6.png)

And here is my config:

```elisp
(use-package zoom
  :ensure t
  :hook (after-init . zoom-mode)
  :config (setq zoom-size '(0.618 . 0.618))
  
  (add-to-list 'zoom-ignored-major-modes 'helm-major-mode)
  (add-to-list 'zoom-ignored-major-modes 'ediff-mode)

  ;; Helm
  (add-to-list 'zoom-ignored-buffer-name-regexps " *Minibuf-1*")
  (add-to-list 'zoom-ignored-buffer-name-regexps "^*helm")
  (add-to-list 'zoom-ignored-buffer-name-regexps "*helm org inbuffer*")
  (add-to-list 'zoom-ignored-buffer-name-regexps "*Org Help*")
  
  ;; (add-to-list 'zoom-ignored-buffer-name-regexps "^*.pdf")

  ;; (setq zoom-ignore-predicates (quote ((lambda nil (window-minibuffer-p)))))
  ;; (setq zoom-ignore-predicates '((lambda () (< (count-lines (point-min) (point-max)) 20))))
  
  ;; FIXME
  ;; Ediff
  (add-to-list 'zoom-ignored-buffer-name-regexps "*Diff*")
  (add-to-list 'zoom-ignored-buffer-name-regexps "^*Ediff")
  (add-to-list 'zoom-ignored-buffer-name-regexps "^*Ediff Control Panel*")
  (add-to-list 'zoom-ignored-buffer-name-regexps "^*Ediff Registry*")
  ;; don't resize ediff control panel window.
  (defun my/fix-ediff-size ()
    (with-selected-window (get-buffer-window "*Ediff Control Panel*")
      (setq window-size-fixed t)
      (window-resize (selected-window) (- 5 (window-total-height)) nil t)))
  (add-hook 'ediff-after-setup-windows-hook 'my/fix-ediff-size))
```

My Emacs version: `GNU Emacs 28.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.20, cairo version 1.17.3) of 2020-06-20`
zoom-mode version: `20200106.1204 from MELPA`

**Comment by cyrus-and** [comment]
Hmmm there's nothing in that config that explains such behaviour, please try to reproduce it with `emacs -Q` and Zoom only.

**Comment by cyrus-and** [comment]
Does this also happen with the default Zoom settings?

**Comment by stardiviner** [comment]
test with minimum Emacs config

1.  `emacs -q`
2.  `[M-x package-initialize]`
3.  `[M-x load-library RET zoom]`
4.  `[M-x zoom-mode]` For now, zoom-mode works fine with JavaScript buffers.
5.  Then I load my zoom-mode config. Still works fine. Weird.
6.  At least can confirm problem is not on zoom-mode.

I don't know which method can debug this.

**Comment by cyrus-and** [comment]
Is JavaScript the only mode affected? See if you can narrow down the list of packages and configurations you're using so I can try to reproduce it locally.

Also show me the value of the `pre-redisplay-function` variable in the buggy scenario. For example mine is:

```lisp
#f(advice-wrapper :after #f(advice-wrapper :before ignore redisplay--pre-redisplay-functions)
                  zoom--handler)
```

PS is Iosevka the font used in the above screenshot?

**Comment by stardiviner** [comment]
I'm using font "DejaVu Sans Mono" in screenshot.

The value of `pre-redisplay-function` is:
```
#f(advice-wrapper :after #f(advice-wrapper :before ignore redisplay--pre-redisplay-functions)
                  zoom--handler)
```
Seems same. I try to edebug on `zoom--handler` or `zoom--update` both not working.
I will try to narrow down my Emacs configurations. Thanks for your hints. It's helpful.

**Comment by stardiviner** [comment]
I finally figured out why now, It's caused by `lsp-mode`. When I run command `lsp` in JS buffer or Python buffer, all get same problem. If I have not enabled `lsp`, zoom works fine on those buffer.

**Comment by cyrus-and** [comment]
Weird, I see no problem with LSP in my setup.

**Comment by stardiviner** [comment]
What about enable `lsp-mode` on buffer?

**Comment by cyrus-and** [comment]
Try with the following, make sure to:

- start without any `~/.emacs.d` directory;
- select `ts-ls` when prompted to;
- `i`mport the project when prompted to.

`zoom-issue-30.el`:

```lisp
(package-initialize)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))
(package-refresh-contents)
(package-install 'zoom)
(package-install 'lsp-mode)

(zoom-mode)

(js-mode)
(call-interactively 'lsp)

(split-window-right)
(split-window-right)
(split-window-right)

(run-at-time 1 nil 'other-window 1)
(run-at-time 2 nil 'other-window 1)
(run-at-time 3 nil 'other-window 1)
(run-at-time 4 nil 'other-window 1)
```

Run Emacs with: `emacs -Q zoom-issue-30.el --eval '(eval-buffer)'`

If the issue doesn't show up then tune this snippet in order to make it happen so I can try locally.

**Comment by stardiviner** [comment]
Tested, it's fine with your minimal Emacs config. I thanks for helping me. So close issue now.

**Comment by stardiviner** [comment]
I found a helper extension `bug-hunter` help me do Emacs init bisect debug. I found problem still on `lsp-mode`.
Then I realized the problem maybe is on `lsp-mode-hook`. I then checked out the value of `lsp-mode-hook`.
It has value:
```
(dap-tooltip-activate-mouse-motions-if-enabled lsp-treemacs-sync-mode lsp-ui-mode lsp-enable-which-key-integration)
```
I checkout `dap-tooltip-activate-mouse-motions-if-enabled` source code, found it use `dapt-tooltip-mode`. I toggle it, found this is the real reason caused this issue.

**Comment by stardiviner** [comment]
I digged a little deeper, the root reason is `(set (make-local-variable 'track-mouse) t)`. This line code cause issue.

**Comment by cyrus-and** [comment]
Good job!

I'm not 100% sure but this smells fishy:

```lisp
(set (make-local-variable 'track-mouse) t)
```

An hackish workaround on my side would be to undo that form inside some of the Zoom functions. Try the following and see if it fixes the issue without introducing glitches in LSP:

```lisp
(defun undo-local-track-mouse(&optional ignored)
  (kill-local-variable 'track-mouse))

(advice-add 'zoom--get-frame-snapshot :before 'undo-local-track-mouse)
(advice-add 'zoom--handler :before 'undo-local-track-mouse)
```

**Comment by stardiviner** [comment]
Yes, this workaround works for me too.
Will this fixed in zoom master branch?

**Comment by cyrus-and** [comment]
Good.

> Will this fixed in zoom master branch?

I don't think so, it's a too niche problem which I suspect is due to an incorrect usage of `track-mouse` in [dap-mode](https://github.com/emacs-lsp/dap-mode/). You can easily add that hack to your configuration file, if anything changes I will post an update here. Thanks for your support.

**Comment by cyrus-and** [comment]
@stardiviner @IvanDyachenko FYI this should be kind of fixed now, see #37, yet the [workaround](https://github.com/cyrus-and/zoom/issues/30#issuecomment-651862486) in this issue still produces a better effect since it does not have (I guess) [this](https://github.com/cyrus-and/zoom/pull/37#issuecomment-745470889) problem.

**Comment by stardiviner** [comment]
I see, thanks for tracking down this problem. @cyrus-and  👍🏻
